README.md
# borderland

Application platform powers Loadbalancers, Microservices and API Gateways

<p align="center">
    <img src="https://raw.githubusercontent.com/wiki/ohyo-io/borderland/images/logo_borderland.png" alt="Borderland logo" width="256" />
</p>

This project was started from ugly fun [example](ttps://github.com/PritiKumr/rust-httpd). But this project is not a simple HTTP load balncer. We dont need it. We need to balance our application payload. So let we use somthing like a application bus to communicate with application logic.

Also there standard practice to use external crates(modules,libraries). But lot of them suffer from overprogramming, not clear implementation of specifications and unnecessary extra functionality. So i liked the starter point of project and i would like to use less external code in that project. Because "less code" is mean less errors and faster executon, which is main goal.

## How to run?

Make sure you have `cargo` [installed](https://www.rust-lang.org/en-US/install.html) and run the command `cargo run` to boot up the server.

Visit `locahost:8080` from your browser.

### What can it do now?

1.  **Say hello world** - visit `localhost:8888/api/v1`

2.  **Serve static files** - visit `localhost:8888/files/index.html` - this will serve the `index.html` file from the `www` folder in the repo root. Place any other file inside `www` and they can be served similarly (using the `/files` prefix - this will be configurable by the user in future, just like in Apache and Nginx).

## Roadmap

- [x] HTTP Scheme redirect to HTTPS by design
- [x] Moved to MIO
- [ ] Routing
- [ ] Session handling
- [ ] Datastorage
- [ ] Application Bus
- [ ] Internal application logic

## Changelog

- Started profiling coz move to use MIO doesnt give good results
- HTTP to HTTPS redirect by design
- HTTPS part and threading
- Moved to MIO event based handling
- Detection plain HTTP by magic (similar as unix `file` command)
- Headers dump
- Small fixes to proper static content handling

## Benchmark

```bash
$ab -n 5000 -kc 500 http://127.0.0.1:8000/

...

Requests per second:    14109.94 [#/sec] (mean)
Time per request:       35.436 [ms] (mean)
Time per request:       0.071 [ms] (mean, across all concurrent requests)
```

It's sad (((


CHANGELOG.md
# Release notes

All notable changes to this project will be documented in this file.

## 0.1.0

- HTTP to HTTPS redirect by design
- HTTPS part and threading
- Moved to MIO event based handling
- Detection plain HTTP by magic (similar as unix `file` command)
- Headers dump
- Small fixes to proper static content handling


src/bak.rs
use std::process::Command;

fn request_url(buffer: &[u8]) -> Option<&str> {
    let mut headers = [httparse::EMPTY_HEADER; 16];
    let mut req = httparse::Request::new(&mut headers);

    match req.parse(&buffer) {
        Ok(_) => match req.path {
            Some(ref path) => {
                return Some(path);
            }
            None => {
                return None;
            }
        },
        Err(_) => {
            return None;
        }
    }
}

fn build_environmental_variables<'a>(
    command: &'a mut Command,
    meta_variables: Vec<(&'a str, &'a str)>,
) {
    for &tup in meta_variables.iter() {
        command.env(tup.0, tup.1);
    }

    println!("{:?}", command);
}

fn build_cgi_meta_vars<'a>(
    request: &'a httparse::Request,
    client_ip: &'a String,
    script_name: &'a str,
    path_info: &'a str,
) -> Vec<(&'a str, &'a str)> {
    let mut headers = Vec::new();

    for (_idx, &item) in request.headers.iter().enumerate() {
        match &item.name {
            &"Authorization" => headers.push(("AUTH_TYPE", str::from_utf8(&item.value).unwrap())),
            &"Content-Length" => {
                headers.push(("CONTENT_LENGTH", str::from_utf8(&item.value).unwrap()))
            }
            &"Content-Type" => headers.push(("CONTENT_TYPE", str::from_utf8(&item.value).unwrap())),
            &"Host" => {
                let header_value = str::from_utf8(&item.value).unwrap();

                match header_value.find(':') {
                    Some(index) => {
                        headers.push(("SERVER_NAME", &header_value[..(index)]));
                        headers.push(("SERVER_PORT", &header_value[(index + 1)..]));
                    }
                    None => {
                        headers.push(("SERVER_NAME", header_value));
                    }
                }
            }
            _ => {}
        };
    }

    headers.push(("REMOTE_ADDR", &client_ip[..]));
    headers.push(("REMOTE_HOST", &client_ip[..]));

    headers.push(("REQUEST_METHOD", request.method.unwrap()));
    headers.push(("SCRIPT_NAME", script_name));

    match path_info.find('?') {
        Some(index) => {
            headers.push(("PATH_INFO", &path_info[..(index)]));
            headers.push(("QUERY_STRING", &path_info[(index + 1)..]));
        }
        None => {
            headers.push(("PATH_INFO", path_info));
        }
    };

    headers.push(("SERVER_PROTOCOL", "HTTP 1.1"));
    headers.push(("SERVER_SOFTWARE", "rust-httpd 0.1"));

    return headers;
}

fn handle_cgi_script(
    request: httparse::Request,
    mut stream: TcpStream,
    client_addr: SocketAddr,
    req_path: &str,
) {
    let path_components: Vec<&str> = req_path.splitn(2, "/").collect();
    let default_path = "/";
    let (script_name, path_info) = (
        path_components.get(0).unwrap(),
        path_components.get(1).unwrap_or(&default_path),
    );

    let client_ip = client_addr.ip().to_string();

    let meta_variables = build_cgi_meta_vars(&request, &client_ip, script_name, path_info);

    let mut command = Command::new(format!("cgi/{}", script_name));

    println!("{:?}", &meta_variables);
    build_environmental_variables(&mut command, meta_variables);

    match command.output() {
        Ok(output) => {
            if output.status.success() {
                stream.write(&output.stdout).expect("Command failed");
            } else {
                stream.write(&output.stderr).expect("Stderr");
            }
        }
        Err(_) => {
            respond_error(stream);
        }
    }
}


src/lib.rs
#[macro_use]
extern crate log;
extern crate env_logger;

extern crate byteorder;
extern crate bytes;
extern crate httparse;
extern crate mio;
extern crate net2;
extern crate slab;

use std::fmt;
use std::num::ParseIntError;
use std::str::FromStr;

mod connection;
mod route;
mod router;
mod server;

pub use self::connection::Connection;
pub use self::route::{Handler, Matcher, ReadWrite, Route};
pub use self::router::Router;
pub use self::server::Server;

pub enum Method {
    OPTIONS,
    GET,
    POST,
    PUT,
    DELETE,
    HEAD,
    TRACE,
    CONNECT,
    PATCH,
    Extension(String),
}

impl FromStr for Method {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.len() {
            3 => match s.as_ref() {
                "GET" => Ok(Method::GET),
                "PUT" => Ok(Method::PUT),
                _ => Ok(Method::Extension(s.to_string())),
            },
            4 => match s.as_ref() {
                "POST" => Ok(Method::POST),
                "HEAD" => Ok(Method::HEAD),
                _ => Ok(Method::Extension(s.to_string())),
            },
            5 => match s.as_ref() {
                "PATCH" => Ok(Method::PATCH),
                "TRACE" => Ok(Method::TRACE),
                _ => Ok(Method::Extension(s.to_string())),
            },
            6 => match s.as_ref() {
                "DELETE" => Ok(Method::DELETE),
                _ => Ok(Method::Extension(s.to_string())),
            },
            7 => match s.as_ref() {
                "OPTIONS" => Ok(Method::OPTIONS),
                "CONNECT" => Ok(Method::CONNECT),
                _ => Ok(Method::Extension(s.to_string())),
            },
            _ => Ok(Method::Extension(s.to_string())),
        }
    }
}

impl fmt::Debug for Method {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "GET")
    }
}


src/router.rs
use super::{Handler, Matcher, Method, ReadWrite, Route};

extern crate conduit_mime_types;
extern crate httparse;
extern crate openssl;

use std::fs::File;
use std::io::{BufRead, BufReader, Read, Write};
use std::net::SocketAddr;
use std::path::Path;
use std::str;
use std::str::FromStr;

#[allow(dead_code)]
pub struct Router {
    options_routes: Vec<Route>,
    get_routes: Vec<Route>,
    post_routes: Vec<Route>,
    put_routes: Vec<Route>,
    delete_routes: Vec<Route>,
    head_routes: Vec<Route>,
    trace_routes: Vec<Route>,
    connect_routes: Vec<Route>,
    patch_routes: Vec<Route>,
    mime_types: conduit_mime_types::Types,
}

impl Router {
    pub fn new() -> Router {
        let _mime_types = conduit_mime_types::Types::new().unwrap();

        Router {
            options_routes: vec![],
            get_routes: vec![],
            post_routes: vec![],
            put_routes: vec![],
            delete_routes: vec![],
            head_routes: vec![],
            trace_routes: vec![],
            connect_routes: vec![],
            patch_routes: vec![],
            mime_types: _mime_types,
        }
    }

    fn respond_hello_world<W: Write>(&self, mut stream: W) {
        let response = b"HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<html><body>Hello world</body></html>\r\n";
        stream.write(response).expect("Write failed");
    }

    fn serve_static_file<W: Write>(&self, mut stream: W, path: &str) {
        println!("STATIC {:?}", format!("www/{}", path));
        let mime_type = self.mime_types.mime_for_path(Path::new(path));
        println!("MIME {:?}", mime_type);
        let mut file = match File::open(format!("www/{}", path)) {
            Ok(file) => file,
            Err(_) => File::open("404.html").expect("404.html file missing!"),
        };

        let mut buffer = Vec::new();
        file.read_to_end(&mut buffer).expect("Read failed");

        println!("\nPATH: {} LENGTH: {}\n", path, buffer.len());

        let response = format!("HTTP/1.1 200 OK\r\nContent-Type: {}\r\n\r\n", mime_type);
        stream.write(response.as_bytes()).expect("Write failed");
        stream.write_all(&buffer).expect("Write failed");
    }

    fn respond_error<W: Write>(&self, mut stream: W) {
        let response = b"HTTP/1.1 500 Internal Server Error\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<html><body>500 - Server Error</body></html>\r\n";
        stream.write(response).expect("Write failed");
    }

    fn respond_file_not_found<W: Write>(&self, mut stream: W) {
        let response = b"HTTP/1.1 404 File Not Found\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<html><body>404 - File Not Found</body></html>\r\n";
        stream.write(response).expect("Write failed");
    }

    pub fn options(mut self, _matcher: Box<Matcher>, handler: Handler) -> Router {
        // path: &str
        // let mut regex = "^".to_string();
        // regex.push_str(path);
        // regex.push_str("$");
        // Path { matcher: Regex::new(&regex).unwrap() }
        self.options_routes.push(Route::new(/*matcher,*/ handler));
        self
    }

    fn read_request_head<T: Read>(&self, stream: T) -> Vec<u8> {
        let mut reader = BufReader::new(stream);
        let mut buff = Vec::new();
        let mut read_bytes = reader
            .read_until(b'\n', &mut buff)
            .expect("reading from stream won't fail");

        while read_bytes > 0 {
            read_bytes = reader.read_until(b'\n', &mut buff).unwrap();
            if read_bytes == 2 && &buff[(buff.len() - 2)..] == b"\r\n" {
                break;
            }
        }
        return buff;
    }

    // #[allow(dead_code)]
    pub fn to_https_scheme<T: ReadWrite>(&self, mut stream: T, _client_addr: SocketAddr) {
        let request_bytes = self.read_request_head(Read::by_ref(&mut stream));
        let mut headers = [httparse::EMPTY_HEADER; 16];
        let mut req = httparse::Request::new(&mut headers);
        let _ = req.parse(&request_bytes);

        println!("HTTP VERSION HTTP 1.{:?}", req.version.unwrap());

        let host = match req.headers.iter().find(|&&header| header.name == "Host") {
            Some(header) => str::from_utf8(header.value).unwrap(),
            None => "",
        };

        let method = Method::from_str(req.method.unwrap());
        let path = req.path.unwrap();

        println!("HOST {:?} {:?} {}\n", host, method.unwrap(), path);

        let response = format!(
            "HTTP/1.1 301 Moved Permanently\r\nLocation: https://{}{}\r\n\r\n",
            "localhost:8443", path
        );
        println!("{}", response);
        stream.write(response.as_bytes()).expect("Write failed");
    }

    pub fn check(&self) {
        println!("ROUTER CHECK");
    }

    pub fn handle<T: ReadWrite>(&self, mut stream: T, _client_addr: SocketAddr) {
        let request_bytes = self.read_request_head(Read::by_ref(&mut stream));
        let mut headers = [httparse::EMPTY_HEADER; 16];
        let mut req = httparse::Request::new(&mut headers);
        req.parse(&request_bytes).unwrap();

        match req.path {
            Some(ref path) => {
                println!("HTTP VERSION HTTP 1.{:?}", req.version.unwrap());

                let host = match req.headers.iter().find(|&&header| header.name == "Host") {
                    Some(header) => str::from_utf8(header.value).unwrap(),
                    None => "",
                };

                let method = Method::from_str(req.method.unwrap());

                println!("HOST {:?} {:?} {}\n", host, method.unwrap(), path);

                for (_i, elem) in req.headers.iter_mut().enumerate() {
                    let s = match str::from_utf8(elem.value) {
                        Ok(v) => v,
                        Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
                    };

                    println!("{:?} {:?}", elem.name, s)
                }

                let _body_length: u32 = match req
                    .headers
                    .iter()
                    .find(|&&header| header.name == "Content-Length")
                {
                    Some(header) => str::from_utf8(header.value).unwrap().parse().unwrap(),
                    None => 0,
                };

                println!("BODY LENGTH {:?}", _body_length);

                // let request_body = read_request_body();

                match req.path {
                    Some(path) => {
                        if path.starts_with("/files") {
                            self.serve_static_file(stream, &path[7..]);
                        } else if path == "/api/v1" {
                            self.respond_hello_world(stream);
                        /*} else if path.starts_with("/cgi") {
                    // DANGER CODE - JUST FOR TESTING
                    handle_cgi_script(req, stream, client_addr, &path[5..]);*/
                        } else {
                            self.respond_file_not_found(stream);
                        }
                    }
                    None => {
                        self.respond_error(stream);
                    }
                };
            }
            None => {
                println!("ROUTER ERROR");
            }
        }
    }
}


src/server.rs
use std::io::{self, ErrorKind};
use std::rc::Rc;

use mio::net::TcpListener;
use mio::unix::UnixReady;
use mio::{Events, Poll, PollOpt, Ready, Token};

use slab;

use connection::Connection;

type Slab<T> = slab::Slab<T, Token>;

pub struct Server {
    // main socket for our server
    sock: TcpListener,

    // token of our server. we keep track of it here instead of doing `const SERVER = Token(_)`.
    token: Token,

    // a list of connections _accepted_ by our server
    conns: Slab<Connection>,
}

const READ_WRITE_CAP: usize = 65536;

impl Server {
    pub fn new(sock: TcpListener) -> Server {
        Server {
            sock: sock,

            // Give our server token a number much larger than our slab capacity. The slab used to
            // track an internal offset, but does not anymore.
            token: Token(10_000_000),

            // We will handle a max of READ_WRITE_CAP connections
            conns: Slab::with_capacity(READ_WRITE_CAP),
        }
    }

    pub fn run(&mut self, poll: &mut Poll) -> io::Result<()> {
        self.register(poll)?;

        info!("SERVER RUN LOOP STARTING...");
        loop {
            let mut events = Events::with_capacity(1024);
            poll.poll(&mut events, None)?;
            for event in events.iter() {
                trace!("EVENT={:?}", event);
                self.ready(poll, event.token(), event.readiness());
            }
        }
    }

    /// Register Server with the poller.
    ///
    /// This keeps the registration details neatly tucked away inside of our implementation.
    pub fn register(&mut self, poll: &mut Poll) -> io::Result<()> {
        poll.register(&self.sock, self.token, Ready::readable(), PollOpt::edge())
            .or_else(|e| {
                error!("Failed to register server {:?}, {:?}", self.token, e);
                Err(e)
            })
    }

    /// Remove a token from the slab
    fn remove_token(&mut self, token: Token) {
        match self.conns.remove(token) {
            Some(_c) => {
                debug!("reset connection; token={:?}", token);
            }
            None => {
                warn!("Unable to remove connection for {:?}", token);
            }
        }
    }

    fn report_slab_size(&mut self) {
        debug!("slab size: {}", self.conns.len());
        // let slsize = self.conns.count();
        // if slsize % 200 == 0 {
        //     println!("now at {} connections", slsize);
        // }
    }

    fn conn_hup(&mut self, poll: &mut Poll, token: Token) -> io::Result<()> {
        debug!("SERVER CONN HUP; tok={:?}", token);
        let res = self.connection(token).hup(poll, true);
        // TODO: hup res ok?
        self.remove_token(token);
        self.report_slab_size();
        res
    }

    fn ready(&mut self, poll: &mut Poll, token: Token, event: Ready) {
        debug!("GOT {:?} EVENT = {:?}", token, event);

        if self.token != token && self.conns.contains(token) == false {
            debug!("Failed to find connection for {:?}", token);
            return;
        }

        let event = UnixReady::from(event);

        if event.is_error() {
            warn!("Error event for {:?}", token);
            self.remove_token(token);
            return;
        }

        if event.is_hup() {
            trace!("HUP EVENT FOR {:?}", token);
            let _ = self.conn_hup(poll, token);
            return;
        }

        let event = Ready::from(event);

        // We never expect a write event for our `Server` token . A write event for any other token
        // should be handed off to that connection.
        if event.is_writable() {
            // println!("### WRITE EVENT FOR {:?}", token);
            assert!(self.token != token, "Received writable event for Server");

            /// Forward a readable event to an established connection.
            ///
            /// Connections are identified by the token provided to us from the poller. Once a read has
            /// finished, push the receive buffer into the all the existing connections so we can
            /// broadcast.
            match self.connection(token).writable(poll) {
                Ok(()) => {
                    let message = b"HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<html><body>Hello world</body></html>\r\n";
                    let rc_message = Rc::new(message.to_vec());
                    self.connection(token)
                        .send_message(rc_message.clone())
                        .unwrap();

                    // while let Some(message) = self.connection(token).readable()? {
                    // let rc_message = Rc::new(message);
                    // self.connection(token).send_message(rc_message.clone())?;
                    // Echo the message too all connected clients.
                    // for c in self.conns.iter_mut() {
                    //     c.send_message(rc_message.clone())?;
                    // }
                    // }

                    // println!("SHOULD HUP FOR HTTP {:?}", token);
                    let _ = self.conn_hup(poll, token);
                    return;
                }
                Err(e) => {
                    warn!("Write event failed for {:?}, {:?}", token, e);
                    self.remove_token(token);
                    return;
                }
            }
        }

        // A read event for our `Server` token means we are establishing a new connection. A read
        // event for any other token should be handed off to that connection.
        if event.is_readable() {
            trace!("### READ EVENT FOR {:?}", token);
            if self.token == token {
                self.accept(poll);
            } else {
                let conn = self.connection(token);
                match conn.readable(poll).unwrap() {
                    // should return empty message and keep it inside connection to reduce data movement as it unused
                    Some(message) => {
                        // println!("GOT MESSAGE {}", String::from_utf8_lossy(&message));
                        // let rc_message = Rc::new(message);
                        // conn.send_message(rc_message.clone()).unwrap();
                    }
                    None => {}
                }
            }
        }

        if self.token != token {
            match self.connection(token).reregister(poll) {
                Ok(()) => {}
                Err(e) => {
                    warn!("Reregister failed {:?}", e);
                    self.remove_token(token);
                    return;
                }
            }
        }
    }

    /// Accept a _new_ client connection.
    ///
    /// The server will keep track of the new connection and forward any events from the poller
    /// to this connection.
    fn accept(&mut self, poll: &mut Poll) {
        debug!("SERVET ACCEPTING NEW SOCKET");

        loop {
            // Log an error if there is no socket, but otherwise move on so we do not tear down the
            // entire server.
            let sock = match self.sock.accept() {
                Ok((sock, _)) => sock,
                Err(e) => {
                    if e.kind() == ErrorKind::WouldBlock {
                        debug!("accept encountered WouldBlock");
                    } else {
                        error!("Failed to accept new socket, {:?}", e);
                    }
                    return;
                }
            };

            let token = match self.conns.vacant_entry() {
                Some(entry) => {
                    let c = Connection::new(sock, entry.index());
                    entry.insert(c).index()
                }
                None => {
                    error!("Failed to insert connection into slab");
                    return;
                }
            };

            debug!("REGISTERING {:?} WITH POLLER", token);
            match self.connection(token).register(poll) {
                Ok(_) => {}
                Err(e) => {
                    error!(
                        "Failed to register {:?} connection with poller, {:?}",
                        token, e
                    );
                    self.remove_token(token);
                }
            }
        }
    }

    /// Find a connection in the slab using the given token.
    ///
    /// This function will panic if the token does not exist. Use self.conns.contains(token)
    /// before using this function.
    fn connection(&mut self, token: Token) -> &mut Connection {
        &mut self.conns[token]
    }
}


src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Read, Write};
use std::net::{SocketAddr, TcpListener, TcpStream};
use std::str;
use std::thread;

extern crate httparse;

fn respond_hello_world(mut stream: TcpStream) {
    let response = b"HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<html><body>Hello world</body></html>\r\n";
    stream.write(response).expect("Write failed");
}

fn serve_static_file(mut stream: TcpStream, path: &str) {
    println!("{:?}", format!("www/{}", path));
    let mut file = match File::open(format!("www/{}", path)) {
        Ok(file) => file,
        Err(_) => File::open("404.html").expect("404.html file missing!"),
    };

    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Read failed");

    let response = b"HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n";
    stream.write(response).expect("Write failed");
    stream.write(&buffer).expect("Write failed");
}

fn respond_error(mut stream: TcpStream) {
    let response = b"HTTP/1.1 500 Internal Server Error\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<html><body>500 - Server Error</body></html>\r\n";
    stream.write(response).expect("Write failed");
}

fn respond_file_not_found(mut stream: TcpStream) {
    let response = b"HTTP/1.1 404 File Not Found\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<html><body>404 - File Not Found</body></html>\r\n";
    stream.write(response).expect("Write failed");
}

fn read_request_head(stream: &TcpStream) -> Vec<u8> {
    let mut reader = BufReader::new(stream);
    let mut buff = Vec::new();
    let mut read_bytes = reader.read_until(b'\n', &mut buff).unwrap();
    while read_bytes > 0 {
        read_bytes = reader.read_until(b'\n', &mut buff).unwrap();
        if read_bytes == 2 && &buff[(buff.len() - 2)..] == b"\r\n" {
            break;
        }
    }
    return buff;
}

/**
 * routing
 */
fn handle_request(stream: TcpStream, _client_addr: SocketAddr) {
    let request_bytes = read_request_head(&stream);
    let mut headers = [httparse::EMPTY_HEADER; 16];
    let mut req = httparse::Request::new(&mut headers);
    let _parsed = req.parse(&request_bytes);

    println!("{:?}", req.method.unwrap());
    println!("{:?}\n", req.path.unwrap());

    for (_i, elem) in req.headers.iter_mut().enumerate() {
        let s = match str::from_utf8(elem.value) {
            Ok(v) => v,
            Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
        };

        println!("{:?} {:?}", elem.name, s)
    }

    let _body_length: u32 = match req
        .headers
        .iter()
        .find(|&&header| header.name == "Content-Length")
    {
        Some(header) => str::from_utf8(header.value).unwrap().parse().unwrap(),
        None => 0,
    };

    println!("BODY LENGTH {:?}", _body_length);

    // let request_body = read_request_body();

    match req.path {
        Some(path) => {
            if path.starts_with("/files") {
                serve_static_file(stream, &path[7..]);
            } else if path == "/api/v1" {
                respond_hello_world(stream);
            /*} else if path.starts_with("/cgi") {
                // DANGER CODE - JUST FOR TESTING
                handle_cgi_script(req, stream, client_addr, &path[5..]);*/
            } else {
                respond_file_not_found(stream);
            }
        }
        None => {
            respond_error(stream);
        }
    };
}

/**
 * main
 */
fn main() {
    let listener = TcpListener::bind("127.0.0.1:8080").unwrap();

    loop {
        match listener.accept() {
            Ok((stream, addr)) => thread::spawn(move || {
                handle_request(stream, addr);
            }),
            Err(e) => thread::spawn(move || println!("Connection failed: {:?}", e)),
        };
    }
}
